src/index.js

```diff
@@ -35,76 +34,85 @@ class CID {
    *
    * The algorithm for argument input is roughly:
    * ```
-   * if (str)
+   * if (cid)
+   *   -> create a copy
+   * else if (str)
    *   if (1st char is on multibase table) -> CID String
    *   else -> bs58 encoded multihash
    * else if (Buffer)
-   *   if (0 or 1) -> CID
+   *   if (1st byte is 0 or 1) -> CID
    *   else -> multihash
    * else if (Number)
    *   -> construct CID by parts
-   *
-   * ..if only JS had traits..
    * ```
    *
    * @param {string|Buffer} version
    * @param {string} [codec]
    * @param {Buffer} [multihash]
+   * @param {string} [multibaseName]
    *
    * @example
-   *
-   * new CID(<version>, <codec>, <multihash>)
+   * new CID(<version>, <codec>, <multihash>, <multibaseName>)
    * new CID(<cidStr>)
    * new CID(<cid.buffer>)
    * new CID(<multihash>)
    * new CID(<bs58 encoded multihash>)
    * new CID(<cid>)
-   *
    */
-  constructor (version, codec, multihash) {
-    let base = null
+  constructor (version, codec, multihash, multibaseName = 'base58btc') {
     if (module.exports.isCID(version)) {
-      let cid = version
+      // version is an exising CID instance
+      const cid = version
       this.version = cid.version
       this.codec = cid.codec
       this.multihash = Buffer.from(cid.multihash)
-      if (cid.base) {
-        this.base = cid.base
-      }
+      this.multibaseName = cid.multibaseName
       return
     }
+
     if (typeof version === 'string') {
-      // base is the base name e.g. 'base32' or false
-      base = multibase.isEncoded(version)
-      if (base) { // CID String (encoded with multibase)
+      // e.g. 'base32' or false
+      const baseName = multibase.isEncoded(version)
+      if (baseName) {
+        // version is a CID String encoded with multibase, so v1
         const cid = multibase.decode(version)
-        version = parseInt(cid.slice(0, 1).toString('hex'), 16)
-        codec = multicodec.getCodec(cid.slice(1))
-        multihash = multicodec.rmPrefix(cid.slice(1))
-      } else { // bs58 string encoded multihash
-        codec = 'dag-pb'
-        multihash = mh.fromB58String(version)
-        version = 0
+        this.version = parseInt(cid.slice(0, 1).toString('hex'), 16)
+        this.codec = multicodec.getCodec(cid.slice(1))
+        this.multihash = multicodec.rmPrefix(cid.slice(1))
+        this.multibaseName = baseName
+      } else {
+        // version is a base58btc string multihash, so v0
+        this.version = 0
+        this.codec = 'dag-pb'
+        this.multihash = mh.fromB58String(version)
+        this.multibaseName = multibaseName
```

Comment:
This is correct and only a matter of taste, so feel free to keep it like it is. I would hard-code it to `base58btc` to make clear that v0 can't have another base.

Reply:
agree!